clippy_lints/src/redundant_async_block.rs

```diff
@@ -0,0 +1,51 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet};
+use rustc_ast::ast::*;
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `async` block that only returns `await` on a future.
+    ///
+    /// ### Why is this bad?
+    /// It is shorter to directly use the future.
+    ///
+    /// ### Example
+    /// ```rust
+    /// await {
+    ///     function_returning_a_future().await
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// function_returning_a_future()
+    /// ```
+    #[clippy::version = "1.69.0"]
+    pub REDUNDANT_ASYNC_BLOCK,
+    complexity,
+    "`async { future.await }` can be replaced by `future`"
+}
+declare_lint_pass!(RedundantAsyncBlock => [REDUNDANT_ASYNC_BLOCK]);
+
+impl EarlyLintPass for RedundantAsyncBlock {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &rustc_ast::Expr) {
+        if expr.span.from_expansion() {
+            return;
+        }
+        if let ExprKind::Async(_, _, block) = &expr.kind && block.stmts.len() == 1 &&
+            let Some(Stmt { kind: StmtKind::Expr(last), .. }) = block.stmts.last() &&
+            let ExprKind::Await(future) = &last.kind
```

Comment:
A check for `future.span.from_expansion()` would be good, and an accompanying test for something like

```rust
macro_rules! await_in_macro {
    ($e:expr) => { std::convert::identity($e).await };
} 

async { await_in_macro!(foo()) }
```

This could come up with `futures::poll` [like here](https://github.com/stratis-storage/stratisd/blob/f20f841756f06c310e5352994ef65d1c40040464/src/engine/structures/lock.rs#L1046) but having our own macro definition means we don't have to rely on [`future`'s implementation](https://docs.rs/futures-util/0.3.26/src/futures_util/async_await/poll.rs.html#17) staying the same

Reply:
You're right, without the test on the `future` itself the proposed solution involved copying code coming from the macro. Done.